Update collection-expressions.md (#7480)
dotnet/csharplang: proposals/csharp-12.0/collection-expressions.md

  * It is assumed that the value of `Count` on a collection will produce that same value as the number of elements when enumerated.
  * The types used in this spec defined in the `System.Collections.Generic` namespace are presumed to be side-effect free.  As such, the compiler can optimize scenarios where such types might be used as intermediary values, but otherwise not be exposed.

>   * It is assumed that a call to some applicable `.AddRange(x)` member on a collection will result in the same final value as iterating over `x` and adding all of its enumerated values individually to the collection with `.Add`.

Author: added this as well @cston

1. Use an existing type that implements that interface.
1. Synthesize a type that implements the interface.

In either case, the type used is allowed to implement a larger set of interfaces than those strictly required.

Synthesized types are free to employ any strategy they want to implement the required interfaces properly.  For example, a synthesized type might inline the elements directly within itself, avoiding the need for additional internal collection allocations.  A synthesized type could also not use any storage whatsoever, opting to compute the values directly.  For example, using `Enumerable.Range(1, 10)` for `[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]`.

#### Non-mutable interface translation
[non-mutable-interface-translation]: #non-mutable-interface-translation

Given a target type or `IEnumerable<T>`, `IReadOnlyCollection<T>`, `IReadOnlyList<T>`, the value generated is allowed to implement more interfaces than required.  For example, implementing the mutable interfaces as well (specifically, implementing `ICollection<T>` or `IList<T>`).  However, in that case:

> 1. The value must return `true` when queried for `ICollection<T>.IsReadOnly`. 
> 1. The value must throw on any call to a mutation method (like `IList<T>.Add`).

Author: ```suggestion
1. The value must return `true` when queried for `ICollection<T>.IsReadOnly`.   This ensures consumers can appropriately tell that the collection it non-mutable, despite implementing the mutable views.
1. The value must throw on any call to a mutation method (like `IList<T>.Add`).  This ensure safety, preventing a non-mutable collection from being accidentally mutated.
```

> 1. The value must return `true` when queried for `ICollection<T>.IsReadOnly`.   This ensures consumers can appropriately tell that the collection it non-mutable, despite implementing the mutable views.

Reviewer: ```suggestion
1. The value must return `true` when queried for `ICollection<T>.IsReadOnly`.   This ensures consumers can appropriately tell that the collection is non-mutable, despite implementing the mutable views.
```

It is recommended that any type that is synthesized implement all these interfaces. This ensures that maximal compatibility with existing libraries, including those that introspect the interfaces implemented by a value in order to light up performance optimizations.

#### Mutable interface translation
[non-mutable-interface-translation]: #non-mutable-interface-translation

Given a target type or `ICollection<T>` or `IList<T>`:

1. The value must return `false` when queried for `ICollection<T>.IsReadOnly`. 

The value generated is allowed to implement more interfaces than required.  Specifically, implementing `IList<T>` even when only targeting `ICollection<T>`.  However, in that case:

> 1. The value support all mutation methods (like `IList<T>.Add`).

Author: ```suggestion
1. The value support all mutation methods (like `IList<T>.RemoveAt`).
```

> 1. The value support all mutation methods (like `IList<T>.RemoveAt`).

Reviewer: ```suggestion
1. The value must support all mutation methods (like `IList<T>.RemoveAt`).
```

    // same assignments as the array translation
    ```

    The translation may use `stackalloc T1[]` or an [*inline array*](https://github.com/dotnet/csharplang/blob/main/proposals/csharp-12.0/inline-arrays.md) rather than `new T1[]` if [*span-safety*](https://github.com/dotnet/csharplang/blob/main/proposals/csharp-7.2/span-safety.md) is maintained.

  * If `T` is some `ReadOnlySpan<T1>`, then the literal is translated the same as for the `Span<T1>` case except that the final result will be that `Span<T1>` [implicitly converted](https://learn.microsoft.com/dotnet/api/system.span-1.op_implicit#system-span-1-op-implicit(system-span((-0)))-system-readonlyspan((-0))) to a `ReadOnlySpan<T1>`.

    The above forms (for arrays and spans) are the base representations of the literal value and are used for the following translation rules.

    * If `T` is some `C<S0, S1, …>` which has a corresponding [create-methods](#create-methods) `B.M<U0, U1, …>()`, then the literal is translated as:

      ```c#
      // Collection literal is passed as is as the single B.M<...>(...) argument
      C<S0, S1, …> __result = B.M<S0, S1, …>([...])
      ```

>       As the `create method` must have an argument type of some instantiated `ReadOnlySpan<T>`, the base translation rule applies there when passing the collection literal to it

Reviewer: ```suggestion
      As the *create method* must have an argument type of some instantiated `ReadOnlySpan<T>`, the translation rule for spans applies when passing the collection expression to the create method.
```

    // same assignments as the array translation
    ```

>     The translation may use `stackalloc T1[]` or an [Inline Array](https://github.com/dotnet/csharplang/blob/main/proposals/csharp-12.0/inline-arrays.md) rather than `new T1[]` if [*span-safety*](https://github.com/dotnet/csharplang/blob/main/proposals/csharp-7.2/span-safety.md) is maintained.

Reviewer: ```suggestion
    The translation may use `stackalloc T1[]` or an [*inline array*](https://github.com/dotnet/csharplang/blob/main/proposals/csharp-12.0/inline-arrays.md) rather than `new T1[]` if [*span-safety*](https://github.com/dotnet/csharplang/blob/main/proposals/csharp-7.2/span-safety.md) is maintained.
```

    // same assignments as the array translation
    ```

    The translation may use `stackalloc T1[]` or an [*inline array*](https://github.com/dotnet/csharplang/blob/main/proposals/csharp-12.0/inline-arrays.md) rather than `new T1[]` if [*span-safety*](https://github.com/dotnet/csharplang/blob/main/proposals/csharp-7.2/span-safety.md) is maintained.

  * If `T` is some `ReadOnlySpan<T1>`, then the literal is translated the same as for the `Span<T1>` case except that the final result will be that `Span<T1>` [implicitly converted](https://learn.microsoft.com/dotnet/api/system.span-1.op_implicit#system-span-1-op-implicit(system-span((-0)))-system-readonlyspan((-0))) to a `ReadOnlySpan<T1>`.

    The above forms (for arrays and spans) are the base representations of the literal value and are used for the following translation rules.

>     * If `T` is some `C<S0, S1, …>` which has a corresponding [create-methods](#create-methods) `B.M<U0, U1, …>()`, then the literal is translated as:

Author: ```suggestion
    * If `T` is some `C<S0, S1, …>` which has a corresponding [create-method](#create-methods) `B.M<U0, U1, …>()`, then the literal is translated as:
```

  </details>

* Can/should the compiler emit Array.Empty for `[]`?  Should we mandate that it does this, to avoid allocations whenever possible?

>     Yes. The compiler should emit Array.Empty<T>() for any case where this is legal and the final result is non-mutable.  For example, targeting `T[]`, `IEnumerable<T>`, `IReadOnlyCollection<T>` or `IReadOnlyList<T>`.  It should not use `Array.Empty<T>` when the target is mutable (`ICollection<T>` or `IList<T>`).

Author: Added this as well.
